src/backend/opencl/kernel/reduce_by_key.hpp

```diff
@@ -0,0 +1,439 @@
+
+            entry.prog = new Program(prog);
+            entry.ker = new Kernel(*entry.prog, "compact");
+
+            addKernelToCache(device, ref_name, entry);
+        }
+
+        NDRange local(threads_x);
+        NDRange global(threads_x * numBlocks);
+
+        auto reduceOp = KernelFunctor<Buffer,
+                                      Buffer, KParam,
+                                      Buffer, KParam,
+                                      Buffer, KParam,
+                                      Buffer, KParam>(*entry.ker);
+
+        reduceOp(EnqueueArgs(getQueue(), global, local),
+                 *reduced_block_sizes,
+                 *keys_out.data, keys_out.info,
+                 *vals_out.data, vals_out.info,
+                         *keys.data, keys.info,
+                         *vals.data, vals.info);
+
+        CL_DEBUG_FINISH(getQueue());
+    }
+
+    template<typename Tk>
+    void launch_test_needs_reduction(cl::Buffer needs_reduction, cl::Buffer needs_boundary,
+                        const Param keys, const int n, const int numBlocks, const int threads_x)
+    {
+        std::string ref_name =
+            std::string("test_needs_reduction_") +
+            std::string(dtype_traits<Tk>::getName()) +
+            std::string("_") +
+            std::to_string(threads_x);
+
+        int device = getActiveDeviceId();
+
+        kc_entry_t entry = kernelCache(device, ref_name);
+
+        if (entry.prog==0 && entry.ker==0) {
+
+            std::ostringstream options;
+            options << " -D Tk=" << dtype_traits<Tk>::getName()
+                    << " -D DIMX=" << threads_x;
+
+            const char *ker_strs[] = {ops_cl, reduce_by_key_needs_reduction_cl};
+            const int   ker_lens[] = {ops_cl_len, reduce_by_key_needs_reduction_cl_len};
+            Program prog;
+            buildProgram(prog, 2, ker_strs, ker_lens, options.str());
+
+            entry.prog = new Program(prog);
+            entry.ker = new Kernel(*entry.prog, "test_needs_reduction");
+
+            addKernelToCache(device, ref_name, entry);
+        }
+
+        NDRange local(threads_x);
+        NDRange global(threads_x * numBlocks);
+
+        auto reduceOp = KernelFunctor<Buffer,
+                                      Buffer,
+                                      Buffer, KParam,
+                                      int>(*entry.ker);
+
+        reduceOp(EnqueueArgs(getQueue(), global, local),
+                 needs_reduction, needs_boundary, *keys.data, keys.info, n);
+
+        CL_DEBUG_FINISH(getQueue());
+    }
+
+    template<typename Ti, typename Tk, typename To, af_op_t op >
+    int reduce_by_key_first(Array<Tk> &keys_out, Array<To> &vals_out,
+                            const Param keys, const Param vals,
+                            bool change_nan, double nanval)
+    {
+        dim4 idims = dim4(keys.info.dims[0], keys.info.dims[1], keys.info.dims[2], keys.info.dims[3]);
```

Comment:
You only use idims[0] once. It would probably be better to use keys.info.dims[0] inline.